feat(RELEASE-1732): conditional system validation in data keys schema

FilipNikolovski opened the pull request:
## Describe your changes
Implement conditional validation for "systems" parameters in data keys JSON schema. Now the systems list contains a list of objects with "systemName" and "dynamic" fields. If the dynamic field is set to true, the system is conditionally required, i.e. we validate its fields, only if the system object is present in the data.

Assisted-by: Cursor

## Relevant Jira

https://issues.redhat.com/browse/RELEASE-1732

## Checklist before requesting a review
- [x] I have marked as draft or added `do not merge` label if there's a dependency PR
  - If you want reviews on your draft PR, you can add reviewers or add the `release-service-maintainers` handle if you are unsure who to tag
- [x] My commit message includes `Signed-off-by: My name <email>`
- [x] I read CONTRIBUTING.MD and [commit formatting](CONTRIBUTING.md#commit-message-formatting-and-standards)
- [x] I have run the README.md generator script in `.github/scripts/readme_generator.sh` and verified the results using `.github/scripts/check_readme.sh`

FilipNikolovski committed:
feat(RELEASE-1732): conditional system validation in data keys schema

Implement conditional validation for "systems" parameters
in data keys JSON schema. Now the systems list contains a list
of objects with "systemName" and "dynamic" fields. If the dynamic
field is set to true, the system is conditionally required, i.e.
we validate its fields, only if the system object is
present in the data.

Assisted-by: Cursor
Signed-off-by: Filip Nikolovski <fnikolov@redhat.com>

seanconroy2021 commented:
> Got it, but now we're changing the "systems" array to be a list of objects instead of strings, is this something that we want?

Yes, that was intentional, as some systems will be mandatory (cdn,releaseNotes) and others will not be (contentGateway), but maybe some of the guys can give their view here.

If we do the above, we will need to pass it in as JSON since Tekton doesn't support the complexity in yaml.
```
- name: systems
  value: |
    [
      { "systemName": "releaseNotes", "dynamic": false }
    ]
```

seanconroy2021 commented:
> I've made an update to the schema, and tested this with a couple of example json files (including the ones you've sent). We also need to update every pipeline that uses the `check-data-keys` right?

Yeah you first need to update check-data-keys task and then update the systems param in check-data-keys task in each pipeline

johnbieren commented on schema/dataKeys.json:
check-data-keys is hardcoded to use the schema from development, so you typically have to update the schema and check-data-keys task in separate PRs

FilipNikolovski replied:
Got it, thanks!

FilipNikolovski replied:
I've opened another PR for the updates in the task https://github.com/konflux-ci/release-service-catalog/pull/1209

FilipNikolovski committed:
feat(RELEASE-1732): update check-data-keys task to follow the new schema

Signed-off-by: Filip Nikolovski <fnikolov@redhat.com>

seanconroy2021 reviewed (dismissed):
LGTM: tested it against a few data.json files I had.

FilipNikolovski commented:
> I would say maybe update the description above to reflect the new way :)

Done!

johnbieren commented:
> > The tests as I can see are not using the schema from development so they fail, would this change still be okay to be merged?
> 
> E2E is currently using your updated schema: https://github.com/FilipNikolovski/release-service-catalog/raw/cad7a71375ca149be61c34d19d2034428acd8432/schema/dataKeys.json
> 
> It’s failing because the schema now expects systems to be an array of objects, but the pipelines haven’t been updated yet to match this structure. I know you have another PR open for the pipeline changes, so I guess the need to be together. The issue is that doing so will cause unit tests to fail, since they run against the development branch not your branch.
> 
> One possible workaround is to temporarily change the default schema reference in the check-data-keys task to point to your updated schema. That way, both E2E and the unit tests would pass. Once everything passes you could change back and merge but maybe their less hackish way.

The pipeline and schema changes should be together. The check-data-keys task changes should be in a separate PR. We should not change the schema to use one from a fork

seanconroy2021 commented:
> The pipeline and schema changes should be together. The check-data-keys task changes should be in a separate PR. We should not change the schema to use one from a fork

Doing the above won't E2E still fail because the current check-data-keys task doesn’t support the new JSON structure, which expects an array of strings, not an array of objects.

FilipNikolovski commented:
> > > The pipeline and schema changes should be together. The check-data-keys task changes should be in a separate PR. We should not change the schema to use one from a fork
> 
> > 
> 
> > Doing the above won't E2E still fail because the current check-data-keys task doesn’t support the new JSON structure, which expects an array of strings, not an array of objects.
> 
> 
> 
> Good point. I think this PR is a first of its kind

In order to avoid this breaking change, we could add a new systems list (call it systemsv2 or whatever) and then update the task and pipelines?

FilipNikolovski commented:
> > > > Temporarily or permanently?
> > 
> > 
> > > 
> > 
> > 
> > > We can do it temporarily, we need to support both the old and new way, until we migrate the pipelines code.
> > 
> > 
> > Yeah, if that is the way forward I think it should be temporary. But, I think every pipeline passes a value for the schema parameter to the check-data-keys task. So, maybe in this one case it would be okay to change the default to your fork, put all of the task and pipelines changes back in this pr (my apologies for incorrectly having you pull them out), then open a PR to put the default back once this one merges?
> 
> This sounds like the better approach, so let's go with this, thanks @johnbieren! I'll just rebase the changes from the other pr onto this one, it's no problem.

Made the updates, moved everything to this PR.

seanconroy2021 commented on pipelines/managed/fbc-release/fbc-release.yaml:
```diff
@@ -52,7 +52,7 @@ spec:
       type: string
       description: The url to the git repo where the release-service-catalog tasks to be used are stored
-      default: https://github.com/konflux-ci/release-service-catalog.git
+      default: https://github.com/FilipNikolovski/release-service-catalog.git
```
So this won't work as you expect. The task Git URL is passed in via the controller from the RPA, so it will never be the default, and you shouldn't need to change it.

seanconroy2021 replied:
Just commenting out the schema from the pipelines and setting the default in check-data keys should be enough.

FilipNikolovski replied:
Ah okay cool, let me revert this.

FilipNikolovski replied:
Done.

seanconroy2021 commented on tasks/managed/check-data-keys/tests/test-check-data-keys.yaml:
```diff
@@ -174,9 +174,7 @@ spec:
-          value:
-            - releaseNotes
-            - cdn
+          value: '[{"systemName": "releaseNotes", "dynamic": false}, {"systemName": "cdn", "dynamic": false}]'
```
I would try to keep it the same as the pipeline. Use a block (|) for defining system param.

FilipNikolovski replied:
Done.

seanconroy2021 commented on tasks/managed/check-data-keys/tests/test-check-data-keys-fail-invalid-releasenotes-key.yaml:
```diff
@@ -140,8 +140,10 @@ spec:
         - name: systems
-          value:
-            - releaseNotes
+          value: |
```
You will need to add two new tests: one for when dynamic is true, so that if the object is not defined, it will pass; and then the opposite when it is false.

FilipNikolovski replied:
I've added two tests to test out the scenarios.

seanconroy2021 commented on tasks/managed/check-data-keys/check-data-keys.yaml:
```diff
@@ -151,10 +150,12 @@ spec:
+        # shellcheck disable=SC2016
+        systemsJSON='$(params.systems)'
 
         jq --argjson systems "$systemsJSON" '.systems += $systems' \
```
Why don't you call the parameter directly inside the JQ?

FilipNikolovski replied:
Done.

seanconroy2021 commented on schema/dataKeys.json:
```diff
@@ -5,20 +5,34 @@
+          "systemName"
         ],
-        "description": "List of systems that are supported to be required in the data"
+        "description": "List of systems to validate dynamically or strictly"
```
I don't think we really need this description, as we normally don't add a description at the root level, just fields inside.

FilipNikolovski replied:
Done.

johnbieren reviewed (commented):
The yaml changes lgtm. I'll let Sean add approval if the schema looks good too. Let's not merge this before today's promotion. It would be much cleaner if it is in the same promotion cycle as the followup PR that reverts the schema param values (hopefully that is clear enough to know what I am talking about.. the thing discussed on the pr)

FilipNikolovski commented:
> The yaml changes lgtm. I'll let Sean add approval if the schema looks good too. Let's not merge this before today's promotion. It would be much cleaner if it is in the same promotion cycle as the followup PR that reverts the schema param values (hopefully that is clear enough to know what I am talking about.. the thing discussed on the pr)

Sounds good :+1:!

FilipNikolovski committed:
test: add tests

Add two tests for check-data-keys
for testing the dynamic field in systems list

Signed-off-by: Filip Nikolovski <fnikolov@redhat.com>

seanconroy2021 reviewed (approved):
LGTM

FilipNikolovski commented:
/retest

FilipNikolovski commented:
/retest
